src/lib/PerceptronPredictor.cc

```diff
@@ -0,0 +1,159 @@
+#include "simeng/PerceptronPredictor.hh"
+
+namespace simeng {
+
+PerceptronPredictor::PerceptronPredictor(ryml::ConstNodeRef config)
+    : btbBits_(config["Branch-Predictor"]["BTB-Tag-Bits"].as<uint64_t>()),
+      globalHistoryLength_(
+          config["Branch-Predictor"]["Global-History-Length"].as<uint64_t>()),
+      rasSize_(config["Branch-Predictor"]["RAS-entries"].as<uint64_t>()) {
+  // Build BTB based on config options
+  uint32_t btbSize = (1 << btbBits_);
+  btb_.resize(btbSize);
+  // Initialise perceptron values with 0 for the global history weights, and 1 for the bias weight;
+  // and intialise the target with 0 (i.e., unknown)
+  for (int i = 0; i < btbSize; i++) {
+    btb_[i].first.assign(globalHistoryLength_, 0);
+    btb_[i].first.push_back(1);
+    btb_[i].second = 0;
+  }
+
+  // Set up training threshold according to empirically determined formula
+  trainingThreshold_ = (int)((1.93 * globalHistoryLength_) + 14);
```

Comment:
if `trainingThreshold_` is a uint64_t, why does the RHS of this need to be interpreted as an int? Should `trainingThreshold_` be of type int, or can we change `(int)` to `(uint64_t)`?